README.md
# id-validator

[简体中文](README.md) | [ENGLISH](README-EN.md)

> China ID number validator. - 中国身份证号验证器。

[![tests](https://github.com/dotecho/id-validator/actions/workflows/tests.yml/badge.svg)](https://github.com/dotecho/id-validator/actions/workflows/tests.yml)
[![codecov](https://codecov.io/gh/guanguans/id-validator/branch/main/graph/badge.svg?token=7X523bI44b)](https://codecov.io/gh/guanguans/id-validator)
[![Go Report Card](https://goreportcard.com/badge/github.com/dotecho/id-validator)](https://goreportcard.com/report/github.com/dotecho/id-validator)
[![GoDoc](https://godoc.org/github.com/dotecho/id-validator?status.svg)](https://godoc.org/github.com/dotecho/id-validator)
[![GitHub release](https://img.shields.io/github/tag/guanguans/id-validator.svg)](https://github.com/dotecho/id-validator/releases)
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/guanguans/id-validator)
![GitHub repo size](https://img.shields.io/github/repo-size/guanguans/id-validator)
![GitHub all releases](https://img.shields.io/github/downloads/guanguans/id-validator/total)
[![GitHub license](https://img.shields.io/github/license/guanguans/id-validator.svg)](https://github.com/dotecho/id-validator/blob/master/LICENSE)

## 功能

* 验证中国身份证号
* 获取身份证号信息
* 升级 15 位身份证号为 18 位
* 伪造符合校验的身份证号

## 环境要求

* Go >= 1.14

## 安装

```shell script
$ go get -u github.com/dotecho/id-validator
```

## 使用

这只是一个快速介绍, 请查看 [GoDoc](https://godoc.org/github.com/dotecho/id-validator) 获得详细信息。

```go
package main

import (
    idvalidator "github.com/dotecho/id-validator"
    "gopkg.in/ffmt.v1"
)

func main() {
    // 验证身份证号合法性
    ffmt.P(idvalidator.IsValid("500154199301135886", true))  // 严格模式验证大陆居民身份证18位
    ffmt.P(idvalidator.IsValid("500154199301135886", false)) // 非严格模式验证大陆居民身份证18位
    ffmt.P(idvalidator.IsValid("11010119900307803X", false)) // 大陆居民身份证末位是X18位
    ffmt.P(idvalidator.IsValid("610104620927690", false))    // 大陆居民身份证15位
    ffmt.P(idvalidator.IsValid("810000199408230021", false)) // 港澳居民居住证18位
    ffmt.P(idvalidator.IsValid("830000199201300022", false)) // 台湾居民居住证18位

    // 获取身份证号信息
    ffmt.P(idvalidator.GetInfo("500154199301135886", true))  // 严格模式获取身份证号信息
    ffmt.P(idvalidator.GetInfo("500154199301135886", false)) // 非严格模式获取身份证号信息
    // []interface {}[
    //     github.com/dotecho/id-validator.IdInfo{          // 身份证号信息
    //         AddressCode: int(500154)                           // 地址码
    //         Abandoned:   int(0)                                // 地址码是否废弃：1为废弃的，0为正在使用的
    //         Address:     string("重庆市市辖区开州区")             // 地址
    //         AddressTree: []string[                             // 省市区三级列表
    //             string("重庆市")                                    // 省
    //             string("市辖区")                                    // 市
    //             string("开州区")                                    // 区
    //         ]
    //         Birthday:      <1993-01-13 00:00:00 +0800 CST>     // 出生日期
    //         Constellation: string("摩羯座")                     // 星座
    //         ChineseZodiac: string("酉鸡")                       // 生肖
    //         Sex:           int(0)                              // 性别：1为男性，0为女性
    //         Length:        int(18)                             // 号码长度
    //         CheckBit:      string("6")                         // 校验码
    //     }
    //     <nil>                                              // 错误信息
    // ]

    // 生成可通过校验的假身份证号
    ffmt.P(idvalidator.FakeId())                                  // 随机生成
    ffmt.P(idvalidator.FakeRequireId(true, "江苏省", "200001", 1)) // 生成出生于2000年1月江苏省的男性居民身份证

    // 15位号码升级为18位
    ffmt.P(idvalidator.UpgradeId("610104620927690"))
    // []interface {}[
    // 	string("610104196209276908") // 升级后号码
    // 	<nil>                        // 错误信息
    // ]
}
```

## 测试

```shell script
$ make test
```

## 变更日志

请参阅 [CHANGELOG](CHANGELOG.md) 获取最近有关更改的更多信息。

## 贡献指南

请参阅 [CONTRIBUTING](.github/CONTRIBUTING.md) 有关详细信息。

## 安全漏洞

请查看[我们的安全政策](../../security/policy)了解如何报告安全漏洞。

## 贡献者

* [guanguans](https://github.com/guanguans)
* [所有贡献者](../../contributors)

## 相关项目

* [jxlwqq/id-validator](https://github.com/jxlwqq/id-validator)，jxlwqq
* [jxlwqq/id-validator.py](https://github.com/jxlwqq/id-validator.py)，jxlwqq
* [mc-zone/IDValidator](https://github.com/mc-zone/IDValidator)，mc-zone
* [renyijiu/id_validator](https://github.com/renyijiu/id_validator)，renyijiu

## 参考资料

* [中华人民共和国公民身份号码](https://zh.wikipedia.org/wiki/中华人民共和国公民身份号码)
* [中华人民共和国民政部：行政区划代码](http://www.mca.gov.cn/article/sj/xzqh/)
* [中华人民共和国行政区划代码历史数据集](https://github.com/jxlwqq/address-code-of-china)
* [国务院办公厅关于印发《港澳台居民居住证申领发放办法》的通知](http://www.gov.cn/zhengce/content/2018-08/19/content_5314865.htm)
* [港澳台居民居住证](https://zh.wikipedia.org/wiki/港澳台居民居住证)

## 协议

MIT 许可证（MIT）。有关更多信息，请参见[协议文件](LICENSE)。


CHANGELOG.md
# Changelog

All notable changes to `guanguans/id-validator` will be documented in this file

## v1.2.9 - 2023-07-04

### What's Changed

- Bump actions/setup-go from 3 to 4 by @dependabot in https://github.com/dotecho/id-validator/pull/32
- Bump github.com/spf13/cast from 1.5.0 to 1.5.1 by @dependabot in https://github.com/dotecho/id-validator/pull/35
- fix panic with wrong format by @shuxiao9058 in https://github.com/dotecho/id-validator/pull/37

### New Contributors

- @shuxiao9058 made their first contribution in https://github.com/dotecho/id-validator/pull/37

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.8...v1.2.9

## v1.2.8 - 2022-09-16

### What's Changed

- Bump github.com/spf13/cast from 1.4.1 to 1.5.0 by @dependabot in https://github.com/dotecho/id-validator/pull/27
- fix: 性别计算错误 ,男性计算为女性 by @jonbenzhang in https://github.com/dotecho/id-validator/pull/29

### New Contributors

- @jonbenzhang made their first contribution in https://github.com/dotecho/id-validator/pull/29

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.7...v1.2.8

## v1.2.7 - 2022-04-30

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.6...v1.2.7

## v1.2.6 - 2022-04-30

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.5...v1.2.6

## v1.2.5 - 2022-04-30

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.4...v1.2.5

## v1.2.4 - 2022-04-30

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.3...v1.2.4

## v1.2.3 - 2022-04-29

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.2...v1.2.3

## v1.2.2 - 2022-04-29

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.1...v1.2.2

## v1.2.1 - 2022-04-29

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.2.0...v1.2.1

## v1.2.0 - 2022-04-29

## What's Changed

- Fix:cannot startup in debug mod in goland by @zzhaoh in https://github.com/dotecho/id-validator/pull/25

## New Contributors

- @zzhaoh made their first contribution in https://github.com/dotecho/id-validator/pull/25

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.11...v1.2.0

## v1.1.11 - 2022-04-29

## What's Changed

- Bump actions/cache from 2.1.7 to 3 by @dependabot in https://github.com/dotecho/id-validator/pull/21
- Bump codecov/codecov-action from 2.1.0 to 3 by @dependabot in https://github.com/dotecho/id-validator/pull/22

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.10...v1.1.11

## v1.1.10 - 2022-03-06

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.9...v1.1.10

## v1.1.9 - 2022-03-04

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.8...v1.1.9

## v1.1.8 - 2022-03-04

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.7...v1.1.8

## v1.1.7 - 2022-03-03

## What's Changed

- Bump actions/setup-go from 2.1.5 to 2.2.0 by @dependabot in https://github.com/dotecho/id-validator/pull/14
- Bump golangci/golangci-lint-action from 2 to 3 by @dependabot in https://github.com/dotecho/id-validator/pull/16
- Bump actions/setup-go from 2.2.0 to 3 by @dependabot in https://github.com/dotecho/id-validator/pull/17
- Bump actions/checkout from 2 to 3 by @dependabot in https://github.com/dotecho/id-validator/pull/18

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.6...v1.1.7

## v1.1.6 - 2022-01-21

## What's Changed

- fix: addressTree's sequence is random by @117503445 in https://github.com/dotecho/id-validator/pull/13

## New Contributors

- @117503445 made their first contribution in https://github.com/dotecho/id-validator/pull/13

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.5...v1.1.6

## v1.1.5 - 2022-01-15

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.4...v1.1.5

## v1.1.4 - 2022-01-14

## What's Changed

- Bump actions/setup-go from 2.1.4 to 2.1.5 by @dependabot in https://github.com/dotecho/id-validator/pull/11

**Full Changelog**: https://github.com/dotecho/id-validator/compare/v1.1.3...v1.1.4

## 1.1.3 - 2021-12-06

- Fix style.
- Add github config files.
- Bump actions/cache from 2.1.6 to 2.1.7.

## 1.1.2 - 2021-09-27

- Bump codecov/codecov-action from 1.0.6 to 2.1.0.
- Bump actions/cache from 1 to 2.1.6.
- Bump actions/setup-go from 1 to 2.1.4.
- Add dependabot config file.
- Update address_code_timeline.go.

## 1.1.1 - 2021-08-30

- Update github action config files.

## 1.1.0 - 2021-05-23

- Add strict mode.
- Add additional address code timeline data.
- Add additional address code timeline data convert script.

## 1.0.3 - 2021-05-22

- Update data to 202012.
- Fix `generatorAddressCode` and `generatorBirthdayCode`.

## 1.0.2 - 2021-04-21

- Fix the ID number with X at the end verification.

## 1.0.1 - 2021-01-27

- Change the access control of some functions.
- Add bench testing.

## 1.0.0 - 2021-01-21

- Initial release.


helper.go
// This file is part of the guanguans/id-validator.
// (c) guanguans <ityaozm@gmail.com>
// This source file is subject to the MIT license that is bundled.

package idvalidator

import (
	"strconv"
	"strings"

	"github.com/dotecho/id-validator/data"
	"github.com/spf13/cast"
)

// 获取地址信息
func getAddressInfo(addressCode string, birthdayCode string, strict bool) map[string]string {
	addressInfo := map[string]string{
		"province": "",
		"city":     "",
		"district": "",
	}

	// 省级信息
	addressInfo["province"] = getAddress(substr(addressCode, 0, 2)+"0000", birthdayCode, strict)

	// 用于判断是否是港澳台居民居住证（8字开头）
	firstCharacter := substr(addressCode, 0, 1)
	// 港澳台居民居住证无市级、县级信息
	if firstCharacter == "8" {
		return addressInfo
	}

	// 市级信息
	addressInfo["city"] = getAddress(substr(addressCode, 0, 4)+"00", birthdayCode, strict)

	// 县级信息
	addressInfo["district"] = getAddress(addressCode, birthdayCode, strict)

	return addressInfo
}

// 获取省市区地址码
func getAddress(addressCode string, birthdayCode string, strict bool) string {
	address := ""
	timeline := data.GetAddressCodeTimeline(cast.ToUint32(addressCode))
	if len(timeline) == 0 {
		// 修复 \d\d\d\d01、\d\d\d\d02、\d\d\d\d11 和 \d\d\d\d20 的历史遗留问题
		// 以上四种地址码，现实身份证真实存在，但民政部历年公布的官方地址码中可能没有查询到
		// 如：440401 450111 等
		// 所以这里需要特殊处理
		// 1980年、1982年版本中，未有制定省辖市市辖区的代码，所有带县的省辖市给予“××××20”的“市区”代码。
		// 1984年版本开始对地级市（前称省辖市）市辖区制定代码，其中“××××01”表示市辖区的汇总码，同时撤销“××××20”的“市区”代码（追溯至1983年）。
		// 1984年版本的市辖区代码分为城区和郊区两类，城区由“××××02”开始排起，郊区由“××××11”开始排起，后来版本已不再采用此方式，已制定的代码继续沿用。
		suffixes := substr("123456", 4, 6)
		switch suffixes {
		case "20":
			address = "市区"
		case "01":
			address = "市辖区"
		case "02":
			address = "城区"
		case "11":
			address = "郊区"
		}

		return address
	}

	year := cast.ToInt(substr(birthdayCode, 0, 4))
	startYear := "0001"
	endYear := "9999"
	for _, val := range timeline {
		if val["start_year"] != "" {
			startYear = val["start_year"]
		}
		if val["end_year"] != "" {
			endYear = val["end_year"]
		}
		if year >= cast.ToInt(startYear) && year <= cast.ToInt(endYear) {
			address = val["address"]
		}
	}

	if address == "" && !strict {
		// 由于较晚申请户口或身份证等原因，导致会出现地址码正式启用于2000年，但实际1999年出生的新生儿，由于晚了一年报户口，导致身份证上的出生年份早于地址码正式启用的年份
		// 由于某些地区的地址码已经废弃，但是实际上在之后的几年依然在使用
		// 这里就不做时间判断了
		address = timeline[0]["address"]
	}

	return address
}

// 获取星座信息
func getConstellation(birthdayCode string) string {
	month, _ := strconv.Atoi(substr(birthdayCode, 4, 6))
	day, _ := strconv.Atoi(substr(birthdayCode, 6, 8))
	startDate := data.Constellation()[month]["start_date"]
	startDay, _ := strconv.Atoi(strings.Split(startDate, "-")[1])
	if day >= startDay {
		return data.Constellation()[month]["name"]
	}

	tmpMonth := month - 1
	if month == 1 {
		tmpMonth = 12
	}

	return data.Constellation()[tmpMonth]["name"]
}

// 获取生肖信息
func getChineseZodiac(birthdayCode string) string {
	// 子鼠
	start := 1900
	end := cast.ToInt(substr(birthdayCode, 0, 4))
	key := (end - start) % 12

	return data.ChineseZodiac()[key]
}

// substr 截取字符串
func substr(source string, start int, end int) string {
	r := []rune(source)
	length := len(r)

	if start < 0 || end > length || start > end {
		return ""
	}

	if start == 0 && end == length {
		return source
	}

	return string(r[start:end])
}


generator.go
// This file is part of the guanguans/id-validator.
// (c) guanguans <ityaozm@gmail.com>
// This source file is subject to the MIT license that is bundled.

package idvalidator

import (
	"fmt"
	"math"
	"math/rand"
	"regexp"
	"strings"
	"time"

	"github.com/dotecho/id-validator/data"
	"github.com/spf13/cast"
)

// 生成Bit码
func generatorCheckBit(body string) string {
	// 位置加权
	var posWeight [19]float64
	for i := 2; i < 19; i++ {
		weight := int(math.Pow(2, float64(i-1))) % 11
		posWeight[i] = float64(weight)
	}

	// 累身份证号body部分与位置加权的积
	var bodySum int
	bodyArray := strings.Split(body, "")
	count := len(bodyArray)
	for i := 0; i < count; i++ {
		bodySum += cast.ToInt(bodyArray[i]) * int(posWeight[18-i])
	}

	// 生成校验码
	checkBit := (12 - (bodySum % 11)) % 11
	if checkBit == 10 {
		return "x"
	}
	return cast.ToString(checkBit)
}

// 生成地址码
func generatorAddressCode(address string) string {
	addressCode := ""
	for code, addressStr := range data.AddressCode() {
		if address == addressStr {
			addressCode = cast.ToString(code)
			break
		}
	}

	classification := addressCodeClassification(addressCode)
	switch classification {
	case "country":
		addressCode = getRandAddressCode("\\d{4}(?)[0-9]{2}$")
	case "province":
		pattern := "^" + substr(addressCode, 0, 2) + "\\d{2}(?)[0-9]{2}$"
		addressCode = getRandAddressCode(pattern)
	case "city":
		pattern := "^" + substr(addressCode, 0, 4) + "(?)[0-9]{2}$"
		addressCode = getRandAddressCode(pattern)
	}

	return addressCode
}

// 地址码分类
func addressCodeClassification(addressCode string) string {
	// 全国
	if addressCode == "" {
		return "country"
	}

	// 港澳台
	if substr(addressCode, 0, 1) == "8" {
		return "special"
	}

	// 省级
	if substr(addressCode, 2, 6) == "0000" {
		return "province"
	}

	// 市级
	if substr(addressCode, 4, 6) == "00" {
		return "city"
	}

	// 县级
	return "district"
}

// 获取随机地址码
func getRandAddressCode(pattern string) string {
	mustCompile := regexp.MustCompile(pattern)
	var keys []string
	for key := range data.AddressCode() {
		keyStr := cast.ToString(key)
		if mustCompile.MatchString(keyStr) && substr(keyStr, 4, 6) != "00" {
			keys = append(keys, keyStr)
		}
	}

	rand.Seed(time.Now().Unix())

	return keys[rand.Intn(len(keys))]
}

// 生成出生日期码
func generatorBirthdayCode(addressCode string, address string, birthday string) string {
	startYear := "0001"
	endYear := "9999"
	year := datePipeHandle(datePad(substr(birthday, 0, 4), "year"), "year")
	month := datePipeHandle(datePad(substr(birthday, 4, 6), "month"), "month")
	day := datePipeHandle(datePad(substr(birthday, 6, 8), "day"), "day")

	if timeline := data.GetAddressCodeTimeline(cast.ToUint32(addressCode)); len(timeline) > 0 {
		for _, val := range timeline {
			if val["address"] == address {
				if val["start_year"] != "" {
					startYear = val["start_year"]
				}
				if val["end_year"] != "" {
					endYear = val["end_year"]
				}
			}
		}
	}

	yearInt := cast.ToInt(year)
	if yearInt < cast.ToInt(startYear) {
		year = startYear
	}
	if yearInt > cast.ToInt(endYear) {
		year = endYear
	}

	birthday = year + month + day
	_, err := time.Parse("20060102", birthday)
	// example: 195578
	if err != nil {
		year = datePad(year, "year")
		month = datePad(month, "month")
		day = datePad(day, "day")
	}

	return year + month + day
}

// 日期处理
func datePipeHandle(date string, category string) string {
	dateInt := cast.ToInt(date)

	switch category {
	case "year":
		nowYear := time.Now().Year()
		rand.Seed(time.Now().Unix())
		if dateInt < 1800 || dateInt > nowYear {
			randDate := rand.Intn(nowYear-1950) + 1950
			date = cast.ToString(randDate)
		}
	case "month":
		if dateInt < 1 || dateInt > 12 {
			randDate := rand.Intn(12-1) + 1
			date = cast.ToString(randDate)
		}

	case "day":
		if dateInt < 1 || dateInt > 31 {
			randDate := rand.Intn(28-1) + 1
			date = cast.ToString(randDate)
		}
	}

	return date
}

// 生成顺序码
func generatorOrderCode(sex int) string {
	rand.Seed(time.Now().Unix())
	orderCode := rand.Intn(999-111) + 111
	if sex != orderCode%2 {
		orderCode--
	}

	return cast.ToString(orderCode)
}

// 日期补全
func datePad(date string, category string) string {
	padLength := 2
	if category == "year" {
		padLength = 4
	}

	return fmt.Sprintf(fmt.Sprintf("%%0%ds", padLength), date)
}


id_validator.go
// This file is part of the guanguans/id-validator.
// (c) guanguans <ityaozm@gmail.com>
// This source file is subject to the MIT license that is bundled.

package idvalidator

import (
	"errors"
	"strconv"
	"time"

	"github.com/dotecho/id-validator/data"
	"github.com/spf13/cast"
)

// IdInfo 身份证信息
type IdInfo struct {
	AddressCode   int
	Abandoned     int
	Address       string
	AddressTree   []string
	Birthday      time.Time
	Constellation string
	ChineseZodiac string
	Sex           int
	Length        int
	CheckBit      string
}

// IsValid 验证身份证号合法性
func IsValid(id string, strict bool) bool {
	code, err := generateCode(id)
	if err != nil {
		return false
	}

	// 检查顺序码、生日码、地址码
	// if !checkOrderCode(code["order"]) || !checkBirthdayCode(code["birthdayCode"]) || !checkAddressCode(code["addressCode"], code["birthdayCode"], strict) {
	if !checkOrderCode(code["order"]) || !checkBirthdayCode(code["birthdayCode"]) {
		return false
	}

	// 15位身份证不含校验码
	if code["type"] == "15" {
		return true
	}

	// 校验码
	return code["checkBit"] == generatorCheckBit(code["body"])
}

// IsLooseValid 宽松验证身份证号合法性
func IsLooseValid(id string) bool {
	return IsValid(id, false)
}

// IsStrictValid 严格验证身份证号合法性
func IsStrictValid(id string) bool {
	return IsValid(id, true)
}

// GetInfo 获取身份证信息
func GetInfo(id string, strict bool) (IdInfo, error) {
	// 验证有效性
	if !IsValid(id, strict) {
		return IdInfo{}, errors.New("invalid ID card number")
	}

	code, _ := generateCode(id)
	addressCode := cast.ToUint32(code["addressCode"])

	// 地址信息
	addressInfo := getAddressInfo(code["addressCode"], code["birthdayCode"], strict)
	addressTree := []string{addressInfo["province"], addressInfo["city"], addressInfo["district"]}

	// 是否废弃
	abandoned := 0
	if data.AddressCode()[addressCode] == "" {
		abandoned = 1
	}

	// 生日
	cst, _ := time.LoadLocation("Asia/Shanghai")
	birthday, _ := time.ParseInLocation("20060102", code["birthdayCode"], cst)

	// 性别
	sex := 1
	order, _ := strconv.Atoi(code["order"])
	if (order % 2) == 0 {
		sex = 0
	}

	// 长度
	length := cast.ToInt(code["type"])

	return IdInfo{
		AddressCode:   int(addressCode),
		Abandoned:     abandoned,
		Address:       addressInfo["province"] + addressInfo["city"] + addressInfo["district"],
		AddressTree:   addressTree,
		Birthday:      birthday,
		Constellation: getConstellation(code["birthdayCode"]),
		ChineseZodiac: getChineseZodiac(code["birthdayCode"]),
		Sex:           sex,
		Length:        length,
		CheckBit:      code["checkBit"],
	}, nil
}

// FakeId 生成假身份证号码
func FakeId() string {
	return FakeRequireId(true, "", "", 0)
}

// FakeRequireId 按要求生成假身份证号码
// isEighteen 是否生成18位号码
// address    省市县三级地区官方全称：如`北京市`、`台湾省`、`香港特别行政区`、`深圳市`、`黄浦区`
// birthday   出生日期：如 `2000`、`198801`、`19990101`
// sex        性别：1为男性，0为女性
func FakeRequireId(isEighteen bool, address string, birthday string, sex int) string {
	// 生成地址码
	addressCode := ""
	if address == "" {
		for i, s := range data.AddressCode() {
			addressCode = cast.ToString(i)
			address = s
			break
		}
	} else {
		addressCode = generatorAddressCode(address)
	}

	// 出生日期码
	birthdayCode := generatorBirthdayCode(addressCode, address, birthday)

	// 生成顺序码
	orderCode := generatorOrderCode(sex)

	if !isEighteen {
		return addressCode + substr(birthdayCode, 2, 8) + orderCode
	}

	body := addressCode + birthdayCode + orderCode

	return body + generatorCheckBit(body)
}

// UpgradeId 15位升级18位号码
func UpgradeId(id string) (string, error) {
	if !IsValid(id, true) {
		return "", errors.New("invalid ID card number")
	}

	code, _ := generateShortCode(id)

	body := code["addressCode"] + code["birthdayCode"] + code["order"]

	return body + generatorCheckBit(body), nil
}


examples/main.go
// This file is part of the guanguans/id-validator.
// (c) guanguans <ityaozm@gmail.com>
// This source file is subject to the MIT license that is bundled.

package main

import (
	idvalidator "github.com/dotecho/id-validator"
	"gopkg.in/ffmt.v1"
)

func main() {
	// 验证身份证号合法性
	ffmt.P(idvalidator.IsValid("500154199301135886", true))  // 严格模式验证大陆居民身份证18位
	ffmt.P(idvalidator.IsValid("500154199301135886", false)) // 非严格模式验证大陆居民身份证18位
	ffmt.P(idvalidator.IsValid("11010119900307803X", false)) // 大陆居民身份证末位是X18位
	ffmt.P(idvalidator.IsValid("610104620927690", false))    // 大陆居民身份证15位
	ffmt.P(idvalidator.IsValid("810000199408230021", false)) // 港澳居民居住证18位
	ffmt.P(idvalidator.IsValid("830000199201300022", false)) // 台湾居民居住证18位

	// 获取身份证号信息
	ffmt.P(idvalidator.GetInfo("500154199301135886", true))  // 严格模式获取身份证号信息
	ffmt.P(idvalidator.GetInfo("500154199301135886", false)) // 非严格模式获取身份证号信息

	// 生成可通过校验的假身份证号
	ffmt.P(idvalidator.FakeId())                                // 随机生成
	ffmt.P(idvalidator.FakeRequireId(true, "江苏省", "200001", 1)) // 生成出生于2000年1月江苏省的男性居民身份证

	// 15位号码升级为18位
	ffmt.P(idvalidator.UpgradeId("610104620927690"))
}
